### PR TITLE
[SessionD] Fix potential redirect flow install bug + light weight refactor on sessiond_main

### PR DESCRIPTION
**Pull request by themarwhal**
## Summary
Two separate commits are in this PR
### Redirect Flow Install
- I was noticing an occasional SessionD crash on CI when running the `TestGyCreditExhaustionRedirect`. From the coredump, it happens when parse a JSON string in RedisClient. I've added a try/catch around the offending lines and added some logs.
- Though I was not able to concretely determine why the JSON parse failed, I've noticed that this crash happens in the DirectoryD response handling thread. This brings some potential synchronization issues since all SessionStore accesses are currently handled in the main sessiond thread. (When a redirect flow is needed, we need to reach out to DirectoryD to get the subscriber's IP address. This I believe is necessary for the CWF case since the IP address is only picked up after the UE completes DHCP.) 
- So this change proposed here is to to have the DirectoryD callback function, that is called after we receive a response, push the session data manipulation onto the main event base. This will guarantee that all of the session store accesses are ordered.
### SessionD Main cleanup
- While I was debugging with GDB, I found that the main function is very long and hard to read. Additionally some of the thread names are not very descriptive. This commit only has cosmetic changes, no real logical change.

## Test Plan
SessionD unit tests
S1AP test
CWF Integ TEst

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/c/session_manager/LocalEnforcer.cpp [hunk 1]
```diff
@@ -442,68 +442,91 @@ static RedirectInformation_AddressType address_type_converter(
...
+  };
+
+  MLOG(MDEBUG) << "Fetching Subscriber IP address from DirectoryD for "
+               << redirect_info.session_id;
+  directoryd_client_->get_directoryd_ip_field(
+      redirect_info.imsi,
+      [this, redirect_info](Status status, DirectoryField resp) {
+        // This call back gets executed in the DirectoryD client thread, but
+        // we want to run the session update logic in the main thread.
+        evb_->runInEventBaseThread([this, redirect_info, status, resp]() {
```
Comment (uri200) [inline lte/gateway/c/session_manager/LocalEnforcer.cpp:542]: what is the implication of delaying the redirection? Do we have to provide any answer to the core that the redirection was done? I understand that while redirection is not done the user will enjoy of some free data>? (which would be fine taking into consideration the lack of real events on sessiond)

File: lte/gateway/c/session_manager/RedisStoreClient.cpp [hunk 2]
```diff
@@ -173,13 +173,19 @@ std::string RedisStoreClient::serialize_session_vec(
...
+  try {
+    folly::dynamic marshaled = folly::parseJson(folly_serialized);
+    for (auto& it : marshaled) {
+      auto stored_session = deserialize_stored_session(it.getString());
+      session_vec.push_back(
+          SessionState::unmarshal(stored_session, *rule_store_));
+    }
+  } catch (std::exception const& e) {
+    // Very rare but we've seen a crash here
+    MLOG(MERROR) << "Exception " << e.what()
```
Comment (uri200) [inline lte/gateway/c/session_manager/RedisStoreClient.cpp:187]: 👍

**Review by uri200 — APPROVED** [review]

File: lte/gateway/c/session_manager/LocalEnforcer.cpp (hunk 1 above)
Comment (themarwhal) [reply lte/gateway/c/session_manager/LocalEnforcer.cpp:542]: @uri200  This is not a delay, but it's more that we want to run all of the logic that touches SessionState/SessionStore in the main thread. This callback function is passed into the DirectoryD client, which means that it's run in DirectoryD's response handling thread. (Different from the main thread) I was wondering if we were seeing any race condition in accessing the session store with these two separate threads accessing the store. So this new logic just moves the handling of the received IP into the main thread.

Comment (themarwhal) [reply lte/gateway/c/session_manager/LocalEnforcer.cpp:542]: Oh, maybe I missed your question. :) The delay should not be that great, as it's only a GRPC exchange with a local DirectoryD service. The redirection install is not triggered by a message from the core, but quota exhaustion. So I suppose waiting for the directoryd response will extend the "free internet" period by a bit, but we already have a 2second error margin due to PipelineD's reporting period anyways.

**Pull request merged**